cfme/utils/auth/__init__.py

```diff
@@ -0,0 +1,178 @@
+from six import iteritems
+import attr
+
+from cfme.configure.configuration.server_settings import (
+    AmazonAuthenticationView, LdapAuthenticationView, LdapsAuthenticationView
+)
+from cfme.exceptions import UnknownProviderType
+from cfme.utils.conf import credentials, auth_data
+
+auth_prov_data = auth_data.get("auth_providers", {})  # setup on module import
+
+USER_TYPES = {
+    'principal': 'User Principal Name',
+    'email': 'E-mail Address',
+    'dn-cn': 'Distinguished Name (CN=<user>)',
+    'dn-uid': 'Distinguished Name (UID=<user>)',
+    'sam': 'SAM Account Name'
+}
+LDAP_PORT = 389
+LDAPS_PORT = 636
+
+
+def auth_provider_types():
+    """Fetch the registered classes from entry_points manageiq.auth_provider_categories"""
+    from pkg_resources import iter_entry_points
+    return {
+        ep.name: ep.resolve()
+        for ep in iter_entry_points('manageiq.auth_provider_types')
+    }
+
+
+def auth_class_from_type(auth_prov_type):
+    """Using the registered auth provider classes, fetch a class by its type key
+
+    Args:
+        auth_prov_type: string key matching a registered type in entry_points
+
+    Raises:
+        UnknownProviderType when the given type isn't registered in entry_points
+    """
+    try:
+        return auth_provider_types()[auth_prov_type]
+    except KeyError:
+        raise UnknownProviderType('Unknown auth provider type: {}'.format(auth_prov_type))
+
+
+def get_auth_crud(auth_prov_key):
+    """Get a BaseAuthProvider derived class with the auth_data.yaml configuration for the key
+
+    Args:
+        auth_prov_key: string key matching one in conf/auth_data.yaml 'auth_providers' dict
+    Raises:
+        ValueError if the yaml type for given key doesn't match auth_type on fetched class
+    """
+    auth_prov_config = auth_prov_data[auth_prov_key].copy()
+    klass = auth_class_from_type(auth_prov_config.get('type'))
+    if auth_prov_config.get('type') != klass.auth_type:
+        raise ValueError('{} must have type "{}"'.format(klass.__name__, klass.auth_type))
+    return klass.from_config(auth_prov_config, auth_prov_key)
+
+
+@attr.s
+class BaseAuthProvider(object):
+    """Base class for authentication provider objects
+
+    Might not be needed
+    """
+    auth_type = None
```

Comment:
attr.ib(default=None) ?

Reply:
I could use attr.ib for this - but I don't want this passed during instantiation so I'd need `init=False` option for attr.ib.  I'll check it out.